File: app/src/main/java/me/liuyun/bjutlgn/util/ThemeHelper.java
package me.liuyun.bjutlgn.util;

import android.app.Activity;
import android.app.Application;
import android.content.Intent;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.annotation.StyleRes;

import java.util.ArrayList;

import java8.util.stream.StreamSupport;
import me.liuyun.bjutlgn.R;
import me.liuyun.bjutlgn.ui.EasterEggActivity;
import me.liuyun.bjutlgn.ui.StatusLockedActivity;

public class ThemeHelper implements Application.ActivityLifecycleCallbacks {
    private ArrayList<Activity> activityList = new ArrayList<>();
    @StyleRes private int currentStyle = 0;
    private static ThemeHelper instance = null;

    private ThemeHelper() {
    }

    public static ThemeHelper getInstance() {
        if (instance == null) {
            instance = new ThemeHelper();
        }
        return instance;
    }

    public int getCurrentStyle() {
        return currentStyle;
    }

    public void init(Application application, @StyleRes int styleRes) {
        application.registerActivityLifecycleCallbacks(this);
        this.currentStyle = styleRes;
    }

    public void setTheme(Activity activity, @StyleRes int styleRes) {
        currentStyle = styleRes;
        PreferenceManager.getDefaultSharedPreferences(activity).edit().putInt("theme", currentStyle).apply();
        StreamSupport.stream(activityList).filter(a -> a != activity).forEach(Activity::recreate);
        Intent intent = new Intent(activity, activity.getClass());
        activity.startActivity(intent);
        activity.overridePendingTransition(R.anim.alpha_in, R.anim.alpha_out);
        activity.finish();
    }

    @Override
    public void onActivityCreated(Activity activity, Bundle bundle) {
        activityList.add(activity);
        if (activity.getClass().getSimpleName().equals(StatusLockedActivity.class.getSimpleName()))
            activity.setTheme(R.style.AppTheme_Dialog);
        else if(activity.getClass().getSimpleName().equals(EasterEggActivity.class.getSimpleName()))
            activity.setTheme(android.R.style.Theme_Wallpaper_NoTitleBar);
        else
            activity.setTheme(currentStyle);
    }

    @Override
    public void onActivityStarted(Activity activity) {

    }

    @Override
    public void onActivityResumed(Activity activity) {

    }

    @Override
    public void onActivityPaused(Activity activity) {

    }

    @Override
    public void onActivityStopped(Activity activity) {

    }

    @Override
    public void onActivitySaveInstanceState(Activity activity, Bundle bundle) {

    }

    @Override
    public void onActivityDestroyed(Activity activity) {
        activityList.remove(activity);
    }
}

File: app/src/main/kotlin/me/liuyun/bjutlgn/util/ThemeHelper.kt
package me.liuyun.bjutlgn.util

import android.app.Activity
import android.app.ActivityManager.TaskDescription
import android.app.Application
import android.content.Context
import android.content.Intent
import android.graphics.BitmapFactory
import android.os.Bundle
import android.preference.PreferenceManager
import androidx.annotation.StyleRes
import android.util.TypedValue
import androidx.core.content.edit
import me.liuyun.bjutlgn.R
import me.liuyun.bjutlgn.ui.EasterEggActivity
import me.liuyun.bjutlgn.ui.StatusLockedActivity
import java.util.*

object ThemeHelper : Application.ActivityLifecycleCallbacks {
    private val activityList = ArrayList<Activity>()
    @StyleRes
    var currentStyle: Int = 0
        private set

    fun init(application: Application, @StyleRes styleRes: Int) {
        application.registerActivityLifecycleCallbacks(this)
        this.currentStyle = styleRes
    }

    @StyleRes
    fun getTheme(context: Context): Int {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        val theme = prefs.getInt("theme", R.style.ThemeBlue)
        if (theme !in ThemeRes.values().map { it.style }) {
            prefs.edit { putInt("theme", R.style.ThemeBlue) }
            return R.style.ThemeBlue
        }
        return theme
    }

    fun setTheme(activity: Activity, @StyleRes styleRes: Int) {
        currentStyle = styleRes
        PreferenceManager.getDefaultSharedPreferences(activity).edit { putInt("theme", currentStyle) }
        activityList.filter { it !== activity }.forEach { it.recreate() }
        val intent = Intent(activity, activity.javaClass)
        activity.startActivity(intent)
        activity.overridePendingTransition(R.anim.alpha_in, R.anim.alpha_out)
        activity.finish()
    }

    fun getThemePrimaryColor(context: Context): Int {
        val value = TypedValue()
        context.theme.resolveAttribute(R.attr.colorPrimary, value, true)
        return value.data
    }

    fun getThemeAccentColor(context: Context): Int {
        val value = TypedValue()
        context.theme.resolveAttribute(R.attr.colorAccent, value, true)
        return value.data
    }

    fun getThemeSecondaryColor(context: Context): Int {
        val value = TypedValue()
        context.theme.resolveAttribute(android.R.attr.colorSecondary, value, true)
        return value.data
    }

    private fun setTaskDescription(activity: Activity) {
        val taskDescription = TaskDescription(activity.getString(R.string.app_name),
                BitmapFactory.decodeResource(activity.resources, R.mipmap.ic_launcher),
                getThemePrimaryColor(activity))
        activity.setTaskDescription(taskDescription)
    }

    override fun onActivityCreated(activity: Activity, bundle: Bundle?) {
        activityList.add(activity)
        when (activity.javaClass.simpleName) {
            StatusLockedActivity::class.java.simpleName -> activity.setTheme(R.style.AppTheme_Dialog)
            EasterEggActivity::class.java.simpleName -> activity.setTheme(android.R.style.Theme_Wallpaper_NoTitleBar)
            else -> activity.setTheme(currentStyle)
        }
        setTaskDescription(activity)
    }

    override fun onActivityStarted(activity: Activity) = Unit
    override fun onActivityResumed(activity: Activity) = Unit
    override fun onActivityPaused(activity: Activity) = Unit
    override fun onActivityStopped(activity: Activity) = Unit
    override fun onActivitySaveInstanceState(activity: Activity, bundle: Bundle?) = Unit

    override fun onActivityDestroyed(activity: Activity) {
        activityList.remove(activity)
    }
}
